app/lib/search/backend.dart

```diff
@@ -60,9 +61,139 @@ void registerPackageIndex(PackageIndex index) =>
+    final lock = GlobalLock.create(
+      'update-search-snapshot',
+      expiration: Duration(minutes: 20),
+    );
+    for (;;) {
+      try {
+        await lock.withClaim((claim) async {
+          await doCreateAndUpdateSnapshot(claim);
+        });
+      } catch (e, st) {
+        _logger.warning('Snapshot update failed.', e, st);
+      }
+      await Future.delayed(Duration(minutes: 1));
+    }
+  }
+
+  @visibleForTesting
+  Future<void> doCreateAndUpdateSnapshot(
+    GlobalLockClaim claim, {
+    Duration sleepDuration = const Duration(minutes: 2),
+  }) async {
+    final firstClaimed = clock.now();
+
+    // The claim will be released after a day, another process may
+    // start to build the snapshot from scratch again.
+    final workUntil = clock.now().add(Duration(days: 1));
+
+    // creating snapshot from scratch
+    final snapshot = SearchSnapshot();
+    Future<void> updatePackage(String package, DateTime? updated) async {
+      // Skip if the last document timestamp is before [updated].
+      // 1-minute window is kept to reduce clock-skew.
+      if (updated != null) {
+        final lastTimestamp = snapshot.documents![package]?.timestamp;
+        if (lastTimestamp != null &&
+            updated.isBefore(lastTimestamp.subtract(Duration(minutes: 1)))) {
+          return;
+        }
+      }
+      // update or remove the document
+      try {
+        final doc = await loadDocument(package);
+        snapshot.add(doc);
+      } on RemovedPackageException catch (_) {
+        snapshot.remove(package);
+      }
+    }
+
+    // initial scan of packages
+    await for (final package in dbService.query<Package>().run()) {
+      if (package.isNotVisible) continue;
+      if (!claim.valid) break;
+      await updatePackage(package.name!, null);
+    }
+    if (!claim.valid) {
+      return;
+    }
+
+    // first complete snapshot, uploading it
+    await _snapshotStorage.uploadDataAsJsonMap(snapshot.toJson());
+    var lastUploadedSnapshotTimestamp = snapshot.updated!;
+
+    // start monitoring
+    var lastQueryStarted = firstClaimed;
+    while (claim.valid) {
+      final now = clock.now().toUtc();
+      if (now.isAfter(workUntil)) {
+        break;
+      }
+
+      lastQueryStarted = now;
+
+      // query updates
+      final recentlyUpdated = await _queryRecentlyUpdated(lastQueryStarted);
+      for (final e in recentlyUpdated.entries) {
+        if (!claim.valid) {
+          break;
+        }
+        await updatePackage(e.key, e.value);
+      }
+
+      if (claim.valid && lastUploadedSnapshotTimestamp != snapshot.updated) {
+        await _snapshotStorage.uploadDataAsJsonMap(snapshot.toJson());
+        lastUploadedSnapshotTimestamp = snapshot.updated!;
+      }
+
+      await Future.delayed(sleepDuration);
+    }
+  }
+
+  Future<Map<String, DateTime>> _queryRecentlyUpdated(
+      DateTime lastQueryStarted) async {
+    final updatedThreshold = lastQueryStarted.subtract(Duration(minutes: 1));
```

Comment:
```suggestion
    final updatedThreshold = lastQueryStarted.subtract(Duration(minutes: 5));
```

Reply:
I think it'd be wise to allow for some drift. My thinking was 15min, but 5 might be okay, 1 is a bit small... I'm just guessing :rofl: